extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java

```diff
@@ -41,20 +55,49 @@ public <T extends InjectableContext.ContextState> CurrentContext<T> create(Class
         return new VertxCurrentContext<>(key);
     }
 
-    /**
-     *
-     * @return an unmodifiable list of used keys
-     */
-    public List<String> keys() {
-        return unmodifiableKeys;
+    public ContextInternal duplicateContextIfContainsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        if (containsAnyCreatedScopeKeys(vertxContext)) {
+            // Duplicate the context, copy the data, remove the request context
+            vertxContext = vertxContext.duplicate();
+            vertxContext.localContextData().putAll(vertxContext.localContextData());
+            keys.forEach(vertxContext.localContextData()::remove);
+            if (requestScopedKeyCreated.get()) {
+                vertxContext.removeLocal(REQUEST_SCOPED_LOCAL_KEY);
+            }
+            VertxContextSafetyToggle.setContextSafe(vertxContext, true);
+        }
+        return vertxContext;
+    }
+
+    private boolean containsAnyCreatedScopeKeys(ContextInternal vertxContext) {
+        boolean requestScopedKeyCreated = this.requestScopedKeyCreated.get();
+        if (requestScopedKeyCreated && vertxContext.getLocal(REQUEST_SCOPED_LOCAL_KEY) != null) {
+            return true;
+        }
+        if (keys.isEmpty()) {
```

Comment:
Not sure what is the most common scenario tbh, but maybe it could help performances to check this as first thing, bringing it at the very beginning of this method?

Reply:
Usually the most common thing is that `requestScopeKeyCreated` is true - it is basically always done AFAIK